File: jzoffer1-20/17.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        
        if(!pRoot1 || !pRoot2) return false;
        //当遇到节点相同的
        if(pRoot1->val==pRoot2->val){
            //判断是否是子结构
            if(isSubtree(pRoot1,pRoot2)){
                return true;
            }
        }
        //判断从左子树开始是否相同
        if(HasSubtree(pRoot1->left,pRoot2)){
            return true;
        }
        //判断右子树是否有此子结构
        if(HasSubtree(pRoot1->right,pRoot2))
        { return true;}
        
        return false;
    }
    
    bool isSubtree(TreeNode* pRoota,TreeNode* pRootb){
        //必须要pRootb在前才可以，为何？因为当他们都为空时，直接返回false了
        if(!pRootb) return true;
        if(!pRoota) return false;
        if(pRoota->val==pRootb->val){
            if(isSubtree(pRoota->left,pRootb->left)&&isSubtree(pRoota->right,pRootb->right)){
                return true;
            }
        }
               return false;
        
    }
};

File: jzoffer1-20/17.js
/* function TreeNode(x) {
    this.val = x;
    this.left = null;
    this.right = null;
} */
function HasSubtree(pRoot1, pRoot2)
{
   if(!pRoot1 || !pRoot2) return false;
    if(pRoot1.val==pRoot2.val){
        if(isSubTree(pRoot1,pRoot2)){
            return true;
        }
    }
    if(HasSubtree(pRoot1.left,pRoot2)){
         return true;
    }
    if(HasSubtree(pRoot1.right,pRoot2)){
        return true; 
    }
    return false;
}
function isSubTree(pRoota,pRootb){
    if(!pRootb) return true;
    if(!pRoota) return false;
    if(pRoota.val==pRootb.val){
        if(isSubTree(pRoota.left,pRootb.left) && isSubTree(pRoota.right,pRootb.right)){
            return true;
        }
    }
    return false;
    
}
/* function TreeNode(x) {
    this.val = x;
    this.left = null;
    this.right = null;
} */
function HasSubtree(pRoot1, pRoot2)
{
    // write code here
    if(pRoot1==null) return false;
    if(pRoot2==null) return false;
    if(pRoot1.val==pRoot2.val)
    {
        if(isSubTree(pRoot1,pRoot2)) return true;
    }
    
      if(HasSubtree(pRoot1.left,pRoot2)) return true;
    
    
        if(HasSubtree(pRoot1.right,pRoot2)) 
        {
    return true;
        }
    return false;
}

function isSubTree(p1,p2){
    if(p2==null) return true;
    if(p1==null) return false;
    if(p1.val==p2.val){
        if(isSubTree(p1.left,p2.left) && isSubTree(p1.right , p2.right)) return true;
    }
    return false;
}